### ISSUE TITLE
type guard strips generic parameter

### ISSUE THREAD

**Issue by vassilyl** [microsoft/pyright#2738]
Use of `TypeGuard[type]` on a generic `Type[T]` changes the type to `type` which is equivalent to `Type[Any]`.

**To Reproduce**

pyright 1.1.197 found 2 errors in the following code

```
from typing import Type, TypeVar
from typing_extensions import TypeGuard


def guard(x) -> TypeGuard[type]:
    return True

T = TypeVar("T")

def unguarded(t: Type[T]) -> T:
    return t()

def guarded(t: Type[T]) -> T:
    if not guard(t):
        raise ValueError()
    return unguarded(t)
```
Pyright messages:
```
  /***.py:18:22 - error: Argument of type "type" cannot be assigned to parameter "t" of type "Type[T@unguarded]" in function "unguarded"
    Type "type" cannot be assigned to type "Type[T@unguarded]" (reportGeneralTypeIssues)
  /***.py:18:12 - error: Expression of type "T@unguarded" cannot be assigned to return type "T@guarded"
    Type "T@unguarded" cannot be assigned to type "T@guarded" (reportGeneralTypeIssues)
```

**Expected behavior**
`Type[T]` should not be narrowed to `Type[Any]`, the code should type-check without errors.

**Comment by erictraut** [comment]
In this issue, the type guard is a red herring. The real issue here is whether the type `type` should be assignable to `Type[T]`.

Here's a simplified sample:
```python
from typing import Type, TypeVar

T = TypeVar("T")

def func1(t: Type[T]) -> T:
    return t()

def func2(t: type):
    reveal_type(func1(t)) # Error: type is not assignable to Type[T@func1]
```

I think the correct behavior is debatable here.

Mypy does not emit an error for the expression `func1(t)`, but the revealed type is `<nothing>` which is clearly incorrect. Pyright emits an error here.

I think you're arguing that `type` should be treated as `Type[Any]` in this case, so it should be assignable to `Type[T]`. It's not clear to me what `T` should be in this case. Should it be `object`? `Any`? Probably the latter.

So, I think our options here are:
1. Stick with the current behavior and disallow `type` to be assigned to `Type[T]`.
2. Treat `type` as the equivalent of `Type[Any]` and allow it to be assigned to `Type[T]`, and then treat `T` as `Any`.

For reasons of consistency, I would lean toward option 2.

**Comment by erictraut** [comment]
I've changed the logic to implement option 2.

**Comment by vassilyl** [comment]
[Python docs](https://docs.python.org/3/library/typing.html#typing.Type) unambiguously support option 2:

> `Type[Any]` is equivalent to `Type` which in turn is equivalent to `type`, which is the root of Python’s metaclass hierarchy.

**Comment by erictraut** [comment]
This is included in pyright 1.1.198, which I just published. It will also be included in the next release of pylance.
